### PR TITLE
feat(sight): capture rustls plaintext for cosh-ng

### LINKED ISSUE #3042: [TestAgent][Nightly][cosh-ng] bug: rustls switch breaks agentsight token capture (CoshNG token_records empty)
### Component

cosh-ng (跨组件,与 sight / agentsight 相关)

### Bug Description

cosh-ng 0.23.0 在 nightly(2026-09-03)上,agentsight 无法再记录 cosh-ng 的 LLM token 使用量。根因是 2026-09-02 的 commit 524dbd6d 把 cosh-core 的 reqwest TLS 后端从 native-tls(OpenSSL)切换到 rustls,导致 cosh-core / cosh-shell 进程不再链接 libssl。agentsight 的 sslsniff eBPF 探针只对 OpenSSL 的 SSL_read / SSL_write / SSL_do_handshake 挂 uprobe,对纯 Rust 的 rustls 进程无法附加(日志大量 "no SSL libraries found in maps"),因此 cosh-ng 的 token_records 完全没有 CoshNG 记录,违反 #2031 token 记录契约与 #2033 prompt-cache 命中可观测性。

该问题由 AgenticOS Nightly 自动化测试在 2026-09-03 nightly(nightly-20260903-020019,anolisa main 6712b3d)如实上报并定位。

### Steps to Reproduce

```shell
# 1. 在 HK ECS 上安装 cosh-ng 0.23.0 与 agentsight,并配置 DashScope key
#    (cosh 从 /root/.bashrc 读 DASHSCOPE_API_KEY 生效)
rpm -q cosh-ng agentsight

# 2. 启动 agentsight trace + serve
systemctl start agentsight

# 3. 启动 cosh TUI 并发起一次真实对话,等待 LLM 回复
tmux new-session -d -s repro 'bash -c "cd /tmp && HOME=/root /usr/bin/cosh"'
tmux send-keys -t repro '你好,测试一下' Enter
# 等待响应(约 30-90s,依赖 DashScope)

# 4. 查询 agentsight token_records 表,观察 cosh-ng 的记录是否落地
python3 -c "import sqlite3; c=sqlite3.connect('file:/var/log/sysak/.agentsight/agentsight.db?mode=ro', uri=True); print(c.execute('SELECT agent, COUNT(*) FROM token_records GROUP BY agent').fetchall())"
# 期望:出现 agent='CoshNG' 的行;实际:只有历史 QwenCode/Hermes/OpenClaw 记录,0 条 CoshNG
```

### Actual Behavior

- cosh 对话成功:cosh-core/sessions 下的 .json 记录了 user "你好,测试一下" 与 assistant qwen-max 真实回复(证明 LLM 调用已成功发起)。
- 但 agentsight 的 token_records 表自 2026-08-30 镜像预热(5 条 QwenCode/Hermes/OpenClaw)后无任何新记录,agent='CoshNG' 行数为 0。
- journalctl -u agentsight 出现 210 条 "[attach_process] ... no SSL libraries found in maps" 警告(cosh 进程因 rustls 无 OpenSSL 符号,SSL uprobe 无法附加)。

### Expected Behavior

cosh-ng 的 LLM 请求应被 agentsight 记录为 agent='CoshNG' 的 token_records(含 input_tokens/output_tokens/cache_read_tokens/cache_creation_tokens),符合 #2031(记录 token 使用)与 #2033(prompt-cache 命中 cache_read_tokens>0)契约。agentsight 内置发现规则 `{"rule": ["*cosh-shell*","*cosh-core*","*cosh-cli*"], "agent_name": "CoshNG"}`(src/agentsight/agentsight.json:79-81)的存在即证明 cosh-ng 应被观测。

### Root Cause

commit 524dbd6d7205(2026-09-02 14:48 +0800, Kailong Zhou)"feat(cosh-ng): switch cosh-core TLS backend to rustls" 将 src/cosh-ng/crates/cosh-core/Cargo.toml 的 reqwest 依赖从 `features = [..., "native-tls"]` 改为 `default-features=false, features=[..., "rustls-tls-native-roots", ...]`,并删除 vendored-openssl feature 与 openssl-libs 运行时依赖。此后 cosh-core 用纯 Rust rustls 建 TLS,进程内存中无 libssl。

agentsight 的 token 抓取(src/agentsight/src/probes/sslsniff.rs)基于 libbpf-rs 对 OpenSSL SSL_read/SSL_write/SSL_do_handshake(+_ex)挂 uprobe 嗅探明文;rustls 不调用这些符号,probe 无法附加。属于跨组件回归:packaging 优化 by-design,但未同步更新 agentsight 观测能力。

### Suggested Fix

需 cosh-ng 与 sight 团队协调,任选其一:

1. 短期回退 rustls:cosh-core Cargo.toml 恢复 native-tls,恢复 openssl 运行时链接(#2980 的 OpenSSL 1.1 安装问题需另寻解法)。
2. 中长期 agentsight 支持 rustls 抓取:对 rustls 握手/record 内部函数挂 uprobe,或改用非 eBPF 的带外 token 上报(复用 #2954 的 DashScope/Bailian native protocol,由 cosh-core 直接把 usage 上报给 agentsight serve)。
3. cosh-core 侧增加显式 token 上报,绕过 TLS 层观测。

```diff
# 示例(方案1,仅供参考 — 具体由团队决策):
- reqwest = { version = "0.12", default-features = false, features = ["blocking", "stream", "json", "http2", "charset", "rustls-tls-native-roots", "system-proxy"] }
+ reqwest = { version = "0.12", features = ["blocking", "stream", "json", "native-tls"] }
```

### Environment

- ALinux 4 ECS(cn-hongkong),kernel 6.6.102-6.2.alnx4.x86_64
- anolisa main: 6712b3d6593e3e82afc5acacd529d93f7fbd3a4c
- cosh-ng 0.23.0-2 / agentsight(daemon active,MemoryMax 350M)

### Relevant Log Output

```
[agentsight::probes::sslsniff:357] [attach_process] pid=NNN: no SSL libraries found in maps
```

### Additional Context

同一根因同时影响兄弟用例 test_agentsight_records_nonzero_tokens(同为 "0 token_records" 签名)。

### PR DESCRIPTION
**Pull request by Daydreamer-Li**
## Why

cosh-ng switched its TLS backend to rustls in 524dbd6d. rustls exports no
`SSL_read`/`SSL_write`, so every uprobe in `sslsniff` became unattachable on
cosh-ng processes — the journal filled with `no SSL libraries found in maps` and
`token_records` held zero `CoshNG` rows, breaking the token-recording contract
from #2031.

While fixing that, a second and independent gap surfaced: an HTTP/2 stream only
completed on the `END_STREAM` flag, which never arrives when a client treats
`data: [DONE]` as the end of the answer and exits before reading the server's
final empty DATA frame. Any HTTP/2 agent hits this, OpenSSL ones included; the
agents that do produce token records today happen to use HTTP/1.1, where a done
marker is already synthesised from the chunked terminator.

## What changed

Two commits, each independently useful:

1. **`fix(sight): complete h2 stream on SSE terminator`** — close an HTTP/2
   stream when a response DATA frame carries the OpenAI-style SSE terminator,
   mirroring the existing HTTP/1.1 behaviour.
2. **`feat(sight): capture rustls plaintext for cosh-ng`** — hook rustls'
   `CommonState::buffer_plaintext` (writes) and
   `CommonState::take_received_plaintext` (reads).

The key decision in (2) is *which* layer to hook. The AEAD layer
(`MessageEncrypter/Decrypter::encrypt`) is also reachable and was tried first,
but its encrypter and decrypter are separate heap objects, so the two directions
report different `ssl_ptr` values, land on different connections, and the HTTP/2
aggregator can never pair a request with its response. Both `CommonState` hooks
take the same `&mut CommonState`, which keeps one connection under one
`ssl_ptr`. Sitting above the record layer additionally means application data
only, one probe per direction instead of one per cipher suite, and no uretprobe
because the plaintext is an argument.

## Related issue

closes #3042

## User / Agent impact

cosh-ng LLM calls become observable: `token_records` and `genai_events` now carry
`agent='CoshNG'` rows with model, provider, token counts, prompt and response.

HTTP/2 agents that previously stalled without `END_STREAM` now complete, so their
token usage is recorded too. No CLI, API, or configuration surface changed.

## Risk and compatibility

- [ ] Public CLI, API, configuration, or documented behavior changed
- [ ] Privileged or security-sensitive behavior changed
- [ ] Cross-component contract changed
- [ ] Migration or rollback guidance is needed

Low risk, but two points deserve review attention.

**The SSE completion path is deliberately narrow.** Only the literal
OpenAI-style terminator counts. No body bytes follow it by protocol, so the
aggregated frames match what a later `END_STREAM` would have produced. Anthropic's
`message_stop` and the DashScope native terminator are *not* recognised, so those
already-working streams keep completing exactly where they do today — a test
pins this. A time-based fallback was rejected because it would alter the timing
semantics of every HTTP/2 connection rather than only the stuck ones.

**The rustls offsets are internals, not a stable API.** `OutboundChunks` is
niche-encoded and `Payload` keeps its pointer and length at the same offsets in
both variants; neither is visible in the type declaration, so both were measured
on a real build with `offset_of!` and cross-checked against the disassembly.
Every BPF-side read is range-checked, so a layout change degrades to capturing
nothing rather than emitting garbage. The 24-byte prologue patterns are likewise
tied to a rustls + toolchain combination and must be re-verified when either
moves; a miss reports the binary as untraceable, which is the pre-existing
behaviour.

Known limits, all intentional and documented in code:

- Detection is keyed on the cosh-ng binary names, so this does not yet cover
  arbitrary rustls processes. Generalising it wants the bounded scan from
  `fix(sight): stream the static SSL pattern scan` first, so an unknown binary is
  not read whole. `find_rustls_offsets` still uses `fs::read`; it only reads the
  cosh-ng binaries (14.6 MiB at most, not the hundreds of megabytes that motivated
  the streaming work), and switching it to `scan_file_patterns` is a follow-up
  once that lands. Verified by cherry-pick that the two changes do not conflict.
- Gather writes are capped: more than 4 chunks, a chunk over 16 KiB, or a total
  over 128 KiB is dropped whole and counted in `rustls_gather_skips`. Dropping
  rather than truncating is deliberate — a short copy desynchronises the frame
  stream and yields garbage instead of partial data.
- x86_64 only, matching agentsight's existing `requires_arch`.

## Validation

Gates, on Alibaba Cloud Linux 4 (kernel 6.6.102, x86_64):

```
cargo fmt --all -- --check                               clean
cargo clippy -p agentsight --all-targets -- -D warnings   clean
cargo test -p agentsight                                  1897 passed, 0 failed
cargo test --test bpf_load -- --ignored                   10 passed (kernel verifier)
```

Ten new tests: five for the SSE completion path (including three that pin the
pre-existing behaviour — plain JSON still completes only on `END_STREAM`, a body
with neither marker stays open, and `message_stop` does not complete early) and
five for rustls detection (both directions located, one-direction still usable,
an ambiguous pattern discarded rather than guessed, absent patterns yielding
`None`, and terminator boundary cases).

End-to-end on a host running cosh-ng 0.23.0 (rustls, `ldd` shows no libssl) with
the service under its normal 350 MiB `MemoryMax`:

```
before:  token_records = []               genai_events = []
after:   token_records = [('CoshNG', 3)]  genai_events = [('CoshNG', 3)]

latest row: model=qwen-plus provider=openai
            input_tokens=4505 output_tokens=34 total_tokens=4539
            http_path=/compatible-mode/v1/chat/completions status=200
            is_sse=1 sse_event_count=9 finish_reasons=["stop"]
            user_query / output_messages / system_instructions all populated

memory: 46 MiB / 350 MiB, oom_kill = 0
```

The frame sequence confirms the intended behaviour: h2 preface, SETTINGS, request
HEADERS, request body as `DATA len=16384` + `DATA END_STREAM len=1828`, then the
state machine advancing `WaitingRequestData` to `RequestComplete` to
`ReceivingResponse`, with every response frame on the same `ConnectionId`.

Not covered: only `cosh-core` was exercised, not `cosh-shell` / `cosh-cli` /
`cosh-gateway`; only DashScope compatible-mode over HTTP/2 and TLS 1.3; and
`cache_read_tokens` stayed 0 because DashScope itself reported `cached_tokens: 0`,
so the #2033 prompt-cache contract could not be confirmed from this run.

## Documentation and rollback

No documentation changes: no CLI, API, or configuration surface moved. The
reasoning that would otherwise live in a design note is in the code — the module
comment in `sslsniff.bpf.c` records why `CommonState` rather than the AEAD layer,
and `response_sse_stream_ended` records why only the literal terminator counts.

Rollback is reverting either commit independently. Neither adds state, a
migration, or a configuration key, so a revert needs no cleanup. Reverting only
the rustls commit leaves the HTTP/2 completion fix in place, which is still
useful on its own.

### CODE REVIEW

File: src/agentsight/src/aggregator/http2.rs [hunk 1]
```diff
@@ -151,6 +151,36 @@ impl Http2StreamState {
...
+    // Both spacings occur in the wild; `data:[DONE]` is legal SSE.
+    const TERMINATORS: [&[u8]; 4] = [
+        b"data: [DONE]",
+        b"data:[DONE]",
+        b"data: [END]",
+        b"data:[END]",
+    ];
+    TERMINATORS
+        .iter()
+        .any(|t| payload.windows(t.len()).any(|w| w == *t))
```
[the remark this answers is not in the document]
Comment (Daydreamer-Li) [reply src/agentsight/src/aggregator/http2.rs]: Fixed. The match now requires that the terminator sits at position 0 or is preceded by `\n`, so a JSON delta whose content happens to contain `"data: [DONE]"` mid-line no longer triggers it. Added an explicit test case that fails against the original code and passes with the fix. Force-pushed in the latest revision.

File: src/agentsight/src/bpf/sslsniff.bpf.c [hunk 2]
```diff
@@ -390,4 +390,277 @@ int BPF_URETPROBE(probe_SSL_do_handshake_exit) {
...
+        return 0;
+    }
+    d->buf_filled = 1;
+    d->buf_size = written;
+    bpf_ringbuf_submit(d, 0);
+    return 0;
+}
+
+SEC("uprobe/rustls_buffer_plaintext")
+int BPF_UPROBE(probe_rustls_write_plaintext, void *conn, void *chunks) {
```
[the remark this answers is not in the document]
Comment (Daydreamer-Li) [reply src/agentsight/src/bpf/sslsniff.bpf.c:612]: Acknowledged as a real gap — this is not a false positive.

A true end-to-end test for this path needs a real rustls process making TLS traffic under a live kernel, which means spinning up cosh-core with a DashScope mock in CI. That is out of scope for this PR but worth tracking.

What is covered today:

- **BPF verifier**: `bpf_load.rs` — all 10 skeleton loading tests pass (including `sslsniff_bpf_loads`), confirming every new BPF program passes the kernel verifier.
- **Byte-pattern detection**: 5 unit tests in `sslsniff.rs` — both directions found, one-direction-only still usable, ambiguous patterns discarded, absent patterns yield `None`.
- **H2 stream completion**: 6 unit tests in `http2.rs` — `[DONE]` closes a stream without `END_STREAM`, plain JSON still only closes on `END_STREAM`, Anthropic `message_stop` does not close early, line-boundary false-positive rejected.
- **Real-machine validation**: 3 DashScope conversations with cosh-ng 0.23.0 (rustls), all 3 produced `token_records` + `genai_events` with correct model, tokens, prompt, and response, under the production 350 MiB `MemoryMax`.

**Review by jfeng18 — APPROVED** [review]
## Deep review — SSE terminator completion + rustls plaintext capture

Two independently useful commits. Both verified on ECS (kernel 6.6.102): fmt ✓ / arch ✓ / clippy ✓ / http2 43/43 ✓ / sslsniff 31/31 ✓.

### Commit 1: `fix(sight): complete h2 stream on SSE terminator`

**The problem**: HTTP/2 stream only closed on END_STREAM. Clients that exit after `[DONE]` never read the server's final empty DATA frame, so the stream sat in ReceivingResponse forever.

**The fix**: close the stream when a response DATA frame carries the OpenAI-style SSE terminator. Deliberately narrow — only `data: [DONE]` / `data: [END]` (both spacings), not Anthropic's `message_stop`.

**Verified**:
- Terminator detection uses `payload.windows(t.len()).any()` — handles embedded terminators in multi-event frames. Correct.
- Anthropic streams are NOT affected — the test `anthropic_terminator_is_left_to_end_stream` pins this. Correct.
- 5 discriminating tests: revert the `|| sse_ended` and the first test fails.

**Edge case**: a terminator split across two DATA frames (e.g., `data: [` in one, `DONE]\n\n` in the next) is not detected. The stream stays open. This is the same limitation as the HTTP/1.1 chunked terminator path, and in practice `[DONE]` arrives as a single SSE event in a single frame. Acceptable.

### Commit 2: `feat(sight): capture rustls plaintext for cosh-ng`

**The problem**: cosh-ng switched to rustls, which exports no SSL_read/SSL_write. Every sslsniff uprobe was unattachable.

**The fix**: hook rustls' `CommonState::buffer_plaintext` (writes) and `CommonState::take_received_plaintext` (reads) via 24-byte prologue scan on stripped binaries.

**Verified**:

- **Layer choice is correct**: CommonState hooks share the same `&mut CommonState` for both directions, so one connection = one `ssl_ptr`. The AEAD layer's separate encrypter/decrypter objects would split the connection and break HTTP/2 request/response correlation.
- **Niche encoding is correctly handled**: `OutboundChunks` tag=0 → Single (ptr@8, len@16), non-NULL → Multiple (chunks@0, n@8, start@16, end@24). The comment explicitly documents the trap of reading Multiple's count from Single's len slot.
- **Gather write concatenation is correct**: chunks are intersected with the [start,end) window and concatenated into a single record. Never truncated — a short copy would desynchronise the HTTP/2 frame parser. Dropped whole + counted in `rustls_gather_skips`.
- **Tier sizing is right**: `RUSTLS_GATHER_TIER = SSL_TIER_MEDIUM = 128 KiB` covers a realistic chat request (18 KiB observed). `RUSTLS_GATHER_CHUNK_MAX = SSL_TIER_SMALL = 16 KiB` covers one HTTP/2 frame.
- **Prologue scan is sound**: 24-byte patterns verified to match exactly once. Ambiguous matches are discarded (not resolved by heuristic) — a misplaced uprobe would sample arbitrary memory as plaintext.
- **Degradation is safe**: a layout change → no pattern match → no capture → no garbage. Every read is range-checked.
- **Detection scope**: keyed on cosh-ng binary names (`cosh-core`, `cosh-shell`, `cosh-cli`, `cosh-gateway`). Generalisation to arbitrary rustls processes is a planned follow-up (needs the bounded scan from the stream-static-SSL PR first).

**Minor observations** (non-blocking):

1. `data: [END]` in the terminator list — which provider sends this? OpenAI uses `[DONE]`. If it's speculative, it might be worth documenting the source.
2. The `rustls_gather_skips` counter is exposed via `pub fn` but there's no API endpoint to read it. Fine for now — it's a coverage gap indicator, not a user-facing metric.

### Verdict

Both commits are correct and well-tested. The rustls capture is a significant enabler for cosh-ng observability.

**Review state: APPROVE**
